[PackageCMO] Don't allow modifying AST

elsh opened the pull request:
Currently not all types are visited in canSerialize* calls, sometimes
resulting in an internal type getting @usableFromInline, which is
incorrect.

For example, for `let q = P() as? Q`, where Q is an internal class
inherting a public class P, Q is not visited in the canSerialize*
checks, thus resulting in `@usableFromInline class Q`; this is not
the intended behavior in the conservative mode used by PackageCMO
as it modifies AST.

To properly fix, instruction visitor needs to be refactored to do
both the "canSerialize" check (that visits all types) and serialize
or update visibility (modify AST in non-conservative modes).

This PR provides a short-term fix that prevents modifying AST, and
also ensures that the generated interfaces with PackageCMO flags
are not affected by the optimization or contain modified AST.

rdar://130292190

elsh commented:
@swift-ci smoke test

elsh commented:
cc @aschwaighofer @nkcsgexi

aschwaighofer reviewed (changes requested):
Are we sure that the current conservative (default) CMO mode does not rely on the current behavior of making all things referenced from an instructions in a function it decides to serialize as usableFromInline?

Would a less risky change (especially for 6.0) be to just disable this behavior only for package CMO?

nkcsgexi commented:
Can we add some tests for this? We could add a textual interface printing test to check that with and without package CMO, we print identical textual interfaces.

aschwaighofer commented:
My comment is not about resilient mode and textual interfaces but the non resilient mode and default CMO - the mode we run today when you don’t use enable-library-evolution.

The failure scenario I am concerned about are projects compile fine today but will fail with a liking error tomorrow after that change.

I don’t think you would find the corner cases by writing a test or two. But rather in the wild.

there was a tail of subtle bugs only found in the wild when Erik introduced CMO.

I would suggest to use the change I had suggested in another forum: check for package CMO and return from the function.

aschwaighofer commented:
And judging from the test failure we are relying on this behavior.

aschwaighofer commented:
(The test tests aggressive cmo AFAICT - which is known to expect to modify the AST. my concern still also holds for conservative cmo)

eeckstein commented on lib/SILOptimizer/IPO/CrossModuleOptimization.cpp:
```diff
@@ -835,14 +835,14 @@ void CrossModuleOptimization::makeDeclUsableFromInline(ValueDecl *decl) {
-    decl->getAttrs().add(attr);
-
     if (everything) {
+      // Mark the nominal type as "usableFromInline".
```
This will degrade performance in aggressive mode significantly, e.g. for the SwiftCompilerSource, which we build in aggressive CMO mode.
You should also allow adding the usableFromInline attribute in aggressive mode.

eeckstein reviewed (changes requested):
This will degrade performance in aggressive mode significantly, e.g. for the SwiftCompilerSource, which we build in aggressive CMO mode.
You should also allow adding the usableFromInline attribute in aggressive mode (`!convervative`)

eeckstein commented:
> Are we sure that the current conservative (default) CMO mode does not rely on the current behavior of making all things referenced from an instructions in a function it decides to serialize as usableFromInline?

It _should_ be fine, because in the conservative mode it only serializes public functions (in which everything should be usable-from-inline anyway).
But you are right that in terms of risk for 6.0 it would be good to narrow down the changed behavior.

eeckstein commented:
As @aschwaighofer told me in a private chat, it's not only a performance issue, the change is wrong.

nkcsgexi commented:
@elsh , Arnold previously suggested this change:
```
--- a/lib/SILOptimizer/IPO/CrossModuleOptimization.cpp
+++ b/lib/SILOptimizer/IPO/CrossModuleOptimization.cpp
@@ -813,6 +813,9 @@ void CrossModuleOptimization::makeFunctionUsableFromInline(SILFunction *function
 
 /// Make a nominal type, including it's context, usable from inline.
 void CrossModuleOptimization::makeDeclUsableFromInline(ValueDecl *decl) {
+  if (isPackageCMOEnabled(M.getSwiftModule()))
+    return;
+
   if (decl->getEffectiveAccess() >= AccessLevel::Package)
     return;
 ```

elsh commented:
> As @aschwaighofer told me in a private chat, it's not only a performance issue, the change is wrong.

Also sync'ed offline. Adding`@usableFromInline` should ideally only happen in non-conservative modes, but some apps might already rely on it, which makes fixing it more tricky and requires some auditing. Added a FIXME comment for now.

elsh commented:
The original change was meant for the `!conservative` modes, but since it might be risky for 6.0, updated to opt out early if in PackageCMO mode.

elsh commented:
@swift-ci smoke test

aschwaighofer commented:
I think we should take this for 6.0 and main in the short term

aschwaighofer commented:
I have put up a medium term alternative here: https://github.com/swiftlang/swift/pull/74695

That alternative shows where the current type checks fail to match behavior of the InstructionVisitor that marks nominal type decls usableFromInline.

IMO we should rely reconcile the two separate ways of checking and then annotating and use an InstructionVisitor for both (long term refactoring).

elsh committed:
[PackageCMO] Don't allow modifying AST.

 Currently not all types are visited in canSerialize* calls, sometimes
 resulting in an internal type getting @usableFromInline, which is
 incorrect.

 For example, for `let q = P() as? Q`, where Q is an internal class
 inherting a public class P, Q is not visited in the canSerialize*
 checks, thus resulting in `@usableFromInline class Q`; this is not
 the intended behavior in the conservative mode used by PackageCMO
 as it modifies AST.

 To properly fix, instruction visitor needs to be refactored to do
 both the "canSerialize" check (that visits all types) and serialize
 or update visibility (modify AST in non-conservative modes).

 This PR provides a short-term fix that prevents modifying AST, and
 also ensures that the generated interfaces with PackageCMO flags
 are not affected by the optimization or contain modified AST.

rdar://130292190

elsh commented:
> I have put up a medium term alternative here: #74695
> 
> That alternative shows where the current type checks fail to match behavior of the InstructionVisitor that marks nominal type decls usableFromInline.
> 
> IMO we should reconcile the two separate ways of checking and annotating and use an InstructionVisitor for both (long term refactoring).

Thanks for the mid-term solution! For this PR, I've updated the test and comments to reflect the issue more accurately.

elsh commented:
@swift-ci smoke test
